ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/remote/streaming/BedrockStreamingHandler.java

```diff
@@ -0,0 +1,435 @@
+            case CONTENT_BLOCK_DELTA:
+                ContentBlockDeltaEvent contentEvent = (ContentBlockDeltaEvent) event;
+
+                // Check if this is a tool use delta
+                if (contentEvent.delta().toolUse() != null) {
+                    log.debug("Tool use delta detected: {}", contentEvent);
+                    accumulateToolInput(contentEvent.delta().toolUse().input(), toolInput, toolInputAccumulator);
+                    return;
+                }
+
+                // Handle regular text content
+                if (contentEvent.delta().text() != null) {
+                    String chunk = contentEvent.delta().text();
+                    sendContentResponse(chunk, false, listener);
+                }
+                break;
+
+            case CONTENT_BLOCK_START:
+                log.debug("Content block started");
+                String[] toolInfo = extractToolInfoFromContentBlockStart(event);
+                toolName.set(toolInfo[0]);
+                toolUseId.set(toolInfo[1]);
+                break;
+
+            case CONTENT_BLOCK_STOP:
+                log.debug("Content block stop");
+                break;
+
+            case MESSAGE_START:
+                log.debug("Message started");
+                break;
+
+            case MESSAGE_STOP:
+                log.debug("Message completed");
+
+                // Check if this is a tool use stop reason
+                if (event.toString().contains(STOP_REASON_TOOL_USE)) {
+                    functionCallInProgress.set(true);
+                    agentExecutionInProgress.set(true);
+                    listener.onResponse(createToolUseResponse(toolName, toolInput, toolUseId));
+                }
+
+                // Close only if no function call or agent execution in progress
+                if (!functionCallInProgress.get() && !agentExecutionInProgress.get()) {
+                    sendCompletionResponse(isStreamClosed, listener);
+                } else {
+                    log.debug("Function call or agent execution in progress - keeping stream open");
+                }
+                break;
+
+            case METADATA:
+                log.debug("Stream metadata: {}", event);
+                break;
+
+            default:
+                log.warn("Unsupported event type: {}", event.sdkEventType());
+                listener.onFailure(new IllegalArgumentException("Unsupported streaming event type: " + event.sdkEventType()));
+                break;
+        }
+    }
+
+    @Override
+    public void handleError(Throwable error, StreamPredictActionListener<MLTaskResponse, ?> listener) {
+        log.error("HTTP streaming error", error);
+        listener.onFailure(new MLException("Fail to execute streaming", error));
+    }
+
+    private MLTaskResponse createToolUseResponse(
+        AtomicReference<String> toolName,
+        AtomicReference<Map<String, Object>> toolInput,
+        AtomicReference<String> toolUseId
```

Comment:
// Validate inputs
    if (toolName == null || toolInput == null || toolUseId == null) {
        throw new IllegalArgumentException("Tool references cannot be null");
    }

This null safety check should be needed.